python/redmonster/physics/zfinder.py

```diff
@@ -4,42 +4,109 @@
 
 # Assumes all templates live in $REDMONSTER_DIR/templates/
 
+def _zchi2(arg) :
+    return zchi2_single_template(**arg)
+
+def _zchi2_no_poly(arg) :
+    return zchi2_single_template_no_poly(**arg)
+
+
+def zchi2_single_template(j,poly_fft, t_fft, t2_fft, data_fft, ivar_fft, pmat_pol, bvec_pol, chi2_0, chi2_null, num_z, npixstep, zminpix, flag_val_neg_model) :
+    npoly=poly_fft.shape[0] # degree+1 of polynomial
+    fftnaxis1=t_fft.shape[0] # number of redshift bins
+    pmat = pmat_pol.copy() # precomputed block for polynomial terms , n.zeros( (npoly+1, npoly+1, fftnaxis1), dtype=float)
+    bvec = bvec_pol.copy() # precomputed vector block for polynomial terms , n.zeros( (npoly+1, fftnaxis1), dtype=float)
+
+    # fill matrix
+    pmat[0,0] = n.fft.ifft(t2_fft * ivar_fft.conj()).real
+    bvec[0]   = n.fft.ifft(t_fft * data_fft.conj()).real
+    for ipos in range(npoly):
+        pmat[ipos+1,0] = pmat[0,ipos+1] = n.fft.ifft(t_fft*poly_fft[ipos].conj()).real
+
+    # solve for each z
+
+    zchi2arr=n.zeros((num_z))
+    zwarning=n.zeros((num_z))
+    for l in n.arange(num_z)*npixstep:
+        try : # try to solve for this redshift
+            f = linalg.solve(pmat[:,:,l+zminpix],bvec[:,l+zminpix])
+
+            zchi2arr[(l//npixstep)] = chi2_0 - n.dot(n.dot(f,pmat[:,:,l+zminpix]),f) # is this true ?????
+            if f[0]<0 :
+                zwarning[(l//npixstep)] = int(zwarning[(l//npixstep)]) | flag_val_neg_model
+                zchi2arr[(l//npixstep)] = chi2_null
+                try:
+                    n.dot(n.dot(f,pmat[:,:,l+zminpix]),f)
+                except Exception as e:
+                    print("Except: %r" % e)
+                    zchi2arr[(l//npixstep)] = chi2_null
+        except : # failure
+            #print "failure"
+            #print sys.exc_info()
+            zchi2arr[(l//npixstep)] = chi2_null
+    return j,zchi2arr,zwarning
+
+def zchi2_single_template_no_poly(j,t_fft, t2_fft, data_fft, ivar_fft, chi2_0, num_z, npixstep, zminpix, flag_val_neg_model) :
+
+    a = n.fft.ifft(t2_fft * ivar_fft.conj()).real
+    b = n.fft.ifft(t_fft * data_fft.conj()).real
+
+    ii = zminpix+n.arange(num_z)*npixstep
+    f = (a[ii]!=0)*b[ii]/(a[ii]+(a[ii]==0))
+    zchi2arr = chi2_0 - a[ii]*f**2 # is this true ?????
+
+    zwarning=n.zeros((num_z))
+    zchi2arr[f<0] = chi2_0
+    zwarning[f<0] = flag_val_neg_model
+    return j,zchi2arr,zwarning
+
+
+
 class ZFinder:
-    
-    def __init__(self, fname=None, group=[0], npoly=None, zmin=None, zmax=None):
+    def __init__(self, fname=None, group=[0], npoly=None, zmin=None, zmax=None, nproc=1):
         self.fname = fname
-        if type(group) == 'list':
+        if type(group) == list:
             self.group = group
+        elif type(group) == int:
+            self.group = [group,]
         else:
             self.group = eval(group)
-        self.npoly = npoly if npoly else 4
+        self.npoly = npoly if npoly is not None else 4
```

Comment:
Note: fixes case for `npoly=0`, which would have been set to `self.npoly = 4` with original code.  This might explain some odd performance vs. npoly seen my Tim and Kyle if they thought they were testing npoly=0 but were actually getting npoly=4.